gc_registry/certificate/services.py

```diff
@@ -304,40 +304,199 @@ def process_certificate_action(
+                certificate_bundle_action.certificate_bundle_percentage
+                * certificate_from_query.bundle_quantity
+                / 100
+            )
+            for certificate_from_query in certificates_from_query
+        ]
+
+    # Only check the bundle quantity if the query on bundle quantity parameter is provided,
+    # otherwise, split the bundle based on the percentage of the total certificates in the bundle
+    for idx, gc_bundle in enumerate(certificates_from_query):
+        gc_bundle = write_session.merge(gc_bundle)
+
+        if certificate_bundle_action.certificate_quantity is not None:
+            if (
+                gc_bundle.bundle_quantity
+                <= certificate_bundle_action.certificate_quantity
+            ):
+                certificates_to_transfer.append(gc_bundle)
+                continue
+
+        chlid_bundle_1, _child_bundle_2 = split_certificate_bundle(
+            gc_bundle,
+            certificates_to_split[idx],
+            write_session,
+            read_session,
+            esdb_client,
+        )
+        if chlid_bundle_1:
+            certificates_to_transfer.append(chlid_bundle_1)
+
+    return certificates_to_transfer
+
+
 def query_certificates(
-    certificate_query: GranularCertificateActionBase, db_read_engine: Session
+    certificate_query: GranularCertificateActionBase,
+    read_session: Session | None = None,
+    write_session: Session | None = None,
 ) -> list[GranularCertificateBundle] | None:
     """Query certificates based on the given filter parameters.
 
+    By default will return read versions of the GC bundles, but if update operations
+    are to be performed on them then passing a write session will override the
+    read session and return instances from the writer database with the associated
+    ActiveUtils methods.
+
+    If no certificates are found with the given query parameters, will return None.
+
     Args:
         certificate_query (GranularCertificateAction): The certificate action
         db_read_engine (Session): The database read session
+        db_write_engine (Session | None): The database write session
 
     Returns:
-        list[GranularCertificateAction]: The list of certificates
+        list[GranularCertificateBundle]: The list of certificates
 
     """
 
+    if (read_session is None) & (write_session is None):
+        logger.error(
+            "Either a read or a write session is required for querying certificates."
+        )
+        return None
+
+    session: Session = read_session if write_session is None else write_session  # type: ignore
+
+    if validate_query(certificate_query) is False:
+        return None
+
     # Query certificates based on the given filter parameters
     stmt = select(GranularCertificateBundle)  # type: ignore
     for query_param, query_value in certificate_query.model_dump().items():
         if (query_param in certificate_query_param_map) & (query_value is not None):
```

Comment:
We should think about limiting the scope of queries to avoid queries bricking infrastructure.

Reply:
> Is there somewhere we are raising a HttpException in the else case?
Address in a separate ticket to split out schemas to make specific to end-point schema